test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PortReserver.cs

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Net;
+using System.Net.NetworkInformation;
+using System.Threading;
+using NuGet.Common;
+
+namespace Dotnet.Integration.Test
+{
+    /// <summary>
+    /// This class allocates ports while ensuring that:
+    /// 1. Ports that are permanently taken (or taken for the duration of the test) are not being attempted to be used.
+    /// 2. Ports are not shared across different tests (but you can allocate two different ports in the same test).
+    /// 
+    /// Gotcha: If another application grabs a port during the test, we have a race condition.
+    /// </summary>
+    [DebuggerDisplay("Port: {PortNumber}, Port count for this app domain: {_appDomainOwnedPorts.Count}")]
+    public class PortReserver : IDisposable
```

Comment:
Thanks for the question!
Actually, there are already two PortReservers in our repo. Both of them are used for testing. The first one has a more general usage and the second one is only used in signing(NuGet.CommandLine.Test)
https://github.com/NuGet/NuGet.Client/blob/dev/test/TestUtilities/Test.Utility/TestServer/PortReserver.cs
https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/PortReserver.cs
There are following differences between the two:
1.APIs are different, the second one has BaseUri property
2.Ways of implementation: The first one tries to new a TcpListenner on that port to check if the port is occupied and uses ConcurrentDictionary to keep the reserved port if it's not occupied. The second one uses Mutex

And I need a third one for signing in Dotnet.integration.Test. I could not use the first one as there are different APIs. I could not use the second one as it's in NuGet.CommandLine.Test which only targets net472, while I need one targets net5.0.

I'll move the second one into Test.Utility as PortReserverForSigning, so that both NuGet.CommandLine.Test and Dotnet.integration.Test could use this one. 